File: android/src/main/java/dev/robingenz/capacitor/badge/BadgePlugin.java
package dev.robingenz.capacitor.badge;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;

@CapacitorPlugin(name = "Badge", permissions = @Permission(strings = {}, alias = "display"))
public class BadgePlugin extends Plugin {

    private Badge implementation;

    @Override
    public void load() {
        implementation = new Badge(getContext());
    }

    @PluginMethod
    public void get(PluginCall call) {
        try {
            int count = implementation.get();
            JSObject ret = new JSObject();
            ret.put("count", count);
            call.resolve(ret);
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void set(PluginCall call) {
        try {
            int count = call.getInt("count", 0);
            implementation.set(count);
            call.resolve();
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void increase(PluginCall call) {
        try {
            implementation.increase();
            call.resolve();
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void decrease(PluginCall call) {
        try {
            implementation.decrease();
            call.resolve();
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void clear(PluginCall call) {
        try {
            implementation.clear();
            call.resolve();
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }
}

File: ios/Plugin/BadgePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(BadgePlugin)
public class BadgePlugin: CAPPlugin {
    private let implementation = Badge()
    
    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        implementation.requestPermissions(completion: { granted, error in
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            call.resolve(["display": granted ? "granted" : "denied"])
        })
    }

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        implementation.checkPermissions(completion: { permission in
            call.resolve([
                "display": permission
            ])
        })
    }

    @objc func get(_ call: CAPPluginCall) {
        let count = implementation.get()
        call.resolve([
            "count": count
        ])
    }
    
    @objc func set(_ call: CAPPluginCall) {
        implementation.requestPermissions(completion: { [weak self] granted, error in
            guard let strongSelf = self else {
                return
            }
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            let count = call.getInt("count") ?? 0
            strongSelf.implementation.set(count: count, completion: {
                call.resolve()
            })
        })
    }
    
    @objc func increase(_ call: CAPPluginCall) {
        implementation.requestPermissions(completion: { [weak self] granted, error in
            guard let strongSelf = self else {
                return
            }
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            strongSelf.implementation.increase(completion: {
                call.resolve()
            })
        })
    }
    
    @objc func decrease(_ call: CAPPluginCall) {
        implementation.requestPermissions(completion: { [weak self] granted, error in
            guard let strongSelf = self else {
                return
            }
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            strongSelf.implementation.decrease(completion: {
                call.resolve()
            })
        })
    }
    
    @objc func clear(_ call: CAPPluginCall) {
        implementation.requestPermissions(completion: { [weak self] granted, error in
            guard let strongSelf = self else {
                return
            }
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            strongSelf.implementation.clear(completion: {
                call.resolve()
            })
        })
    }
}
